File: aoc/2021/aoc202108.hs
import Data.Map((!))
import qualified Data.Map
import Data.Set(Set,fromList,intersection,size)

parse :: String -> [([Set Char],[Set Char])]
parse = p . map fromList . words
  where
    p [] = []
    p items = (take 10 items,drop 11 (take 15 items)) : p (drop 15 items)

is1478 :: Set a -> Bool
is1478 item = size item `elem` [2,4,3,7]

testData :: String
testData = "be cfbegad cbdgef fgaecd cgeb fdcge agebfd fecdb fabcd edb | fdgacbe cefdb cefbgd gcbe\n edbfga begcd cbg gc gcadebf fbgde acbgfd abcde gfcbed gfec | fcgedb cgb dgebacf gc\n fgaebd cg bdaec gdafb agbcfd gdcbef bgcad gfac gcb cdgabef | cg cg fdcagb cbg\n fbegcd cbd adcefb dageb afcb bc aefdc ecdab fgdeca fcdbega | efabcd cedba gadfec cb\n aecbfdg fbg gf bafeg dbefa fcge gcbea fcaegb dgceab fcbdga | gecf egdcabf bgf bfgea\n fgeab ca afcebg bdacfeg cfaedg gcfdb baec bfadeg bafgc acf | gebdcfa ecba ca fadegcb\n dbcfg fgd bdegcaf fgec aegbdf ecdfab fbedc dacgb gdcebf gf | cefg dcbef fcge gbcadfe\n bdfegc cbegaf gecbf dfcage bdacg ed bedf ced adcbefg gebcd | ed bcgafe cdgba cbgef\n egadfb cdbfeg cegd fecab cgb gbdefca cg fgcdab egfdb bfceg | gbdfcae bgc cg cgb\n gcafb gcf dcaebfg ecagb gf abcdeg gaef cafbge fdbac fegbdc | fgae cfgab fg bagce\n"

test :: ()
test
  | (length . filter is1478 . concatMap snd . parse) testData /= 26 = error "a"
  | otherwise = ()

part1 :: IO Int
part1 = fmap (length . filter is1478 . concatMap snd . parse) $ readFile "input/08.txt"

decode :: ([Set Char],[Set Char]) -> String
decode (digits,output) = map (table!) output
  where
    one = head $ filter ((== 2) . size) digits
    seven = head $ filter ((== 3) . size) digits
    four = head $ filter ((== 4) . size) digits
    d235 = filter ((== 5) . size) digits
    d069 = filter ((== 6) . size) digits
    eight = head $ filter ((== 7) . size) digits
    three = head $ filter ((== 2) . size . intersection one) d235
    two = head $ filter ((== 2) . size . intersection four) d235
    five = head $ filter (/= two) $ filter (/= three) d235
    six = head $ filter ((/= one) . intersection one) d069
    nine = head $ filter ((== four) . intersection four) d069
    zero = head $ filter ((/= five) . intersection five) d069
    table = Data.Map.fromList [(zero,'0'),(one,'1'),(two,'2'),(three,'3'),(four,'4'),(five,'5'),(six,'6'),(seven,'7'),(eight,'8'),(nine,'9')]

test2 :: ()
test2
  | (sum . map (read . decode) . parse) testData /= 61229 = error "a"
  | otherwise = ()

part2 :: IO Int
part2 = fmap (sum . map (read . decode) . parse) $ readFile "input/08.txt"

File: aoc/2025/aoc202108.go
package main

import (
	"math/bits"
	"runtime"
)

func init() {
	Register(&aoc202108{
		AOC: AOC{
			Day:           8,
			InputFilename: "../2021/input/08.txt",
			Tests: []Test{
				Test{
					Input: `acedgfb cdfbe gcdfa fbcad dab cefabd cdfgeb eafb cagedb ab | cdfeb fcadb cdfeb cdbaf
`,
					Part1: "",
					Part2: "5353",
				},
				Test{
					Input: `be cfbegad cbdgef fgaecd cgeb fdcge agebfd fecdb fabcd edb | fdgacbe cefdb cefbgd gcbe
edbfga begcd cbg gc gcadebf fbgde acbgfd abcde gfcbed gfec | fcgedb cgb dgebacf gc
fgaebd cg bdaec gdafb agbcfd gdcbef bgcad gfac gcb cdgabef | cg cg fdcagb cbg
fbegcd cbd adcefb dageb afcb bc aefdc ecdab fgdeca fcdbega | efabcd cedba gadfec cb
aecbfdg fbg gf bafeg dbefa fcge gcbea fcaegb dgceab fcbdga | gecf egdcabf bgf bfgea
fgeab ca afcebg bdacfeg cfaedg gcfdb baec bfadeg bafgc acf | gebdcfa ecba ca fadegcb
dbcfg fgd bdegcaf fgec aegbdf ecdfab fbedc dacgb gdcebf gf | cefg dcbef fcge gbcadfe
bdfegc cbegaf gecbf dfcage bdacg ed bedf ced adcbefg gebcd | ed bcgafe cdgba cbgef
egadfb cdbfeg cegd fecab cgb gbdefca cg fgcdab egfdb bfceg | gbdfcae bgc cg cgb
gcafb gcf dcaebfg ecagb gf abcdeg gaef cafbge fdbac fegbdc | fgae cfgab fg bagce
`,
					Part1: "26",
					Part2: "61229",
				},
			},
		},
	})
}

type aoc202108 struct {
	AOC
}

func (aoc *aoc202108) parse(input *Input) Seq[[14]byte] {
	toBits := func(signal string) byte {
		bits := byte(0)
		for _, b := range signal {
			bits |= 1 << (b - 'a')
		}
		return bits
	}
	return func(yield func([14]byte) bool) {
		for {
			entry := [14]byte{}
			for i := range 10 {
				word, ok := input.Word()
				if !ok {
					return
				}
				entry[i] = toBits(word)
			}
			input.Word()
			for i := range 4 {
				word, ok := input.Word()
				if !ok {
					return
				}
				entry[10+i] = toBits(word)
			}
			if !yield(entry) {
				return
			}
		}
	}
}

func (aoc *aoc202108) Part1(input *Input) string {
	in := make(chan [14]byte)
	out := make(chan int)
	for range runtime.NumCPU() {
		go func() {
			result := 0
			for entry := range in {
				for _, b := range entry[10:14] {
					switch bits.OnesCount8(b) {
					case 2, 3, 4, 7:
						result++
					}
				}
			}
			out <- result
		}()
	}
	for entry := range aoc.parse(input) {
		in <- entry
	}
	close(in)
	result := 0
	for range runtime.NumCPU() {
		result += <-out
	}
	return IntResult(result)
}

func (aoc *aoc202108) decode(entry [14]byte) int {
	var one, seven, four, eight byte
	for _, b := range entry[:10] {
		switch bits.OnesCount8(b) {
		case 2:
			one = b
		case 3:
			seven = b
		case 4:
			four = b
		case 7:
			eight = b
		}
	}
	var two, three, five byte
	var zero, six, nine byte
	for _, b := range entry[:10] {
		switch bits.OnesCount8(b) {
		case 5:
			if one&b == one {
				three = b
			} else if bits.OnesCount8(four&b) == 2 {
				two = b
			} else {
				five = b
			}
		case 6:
			if four&b == four {
				nine = b
			} else if one&b == one {
				zero = b
			} else {
				six = b
			}
		}
	}
	digit := func(b byte) int {
		if b == zero {
			return 0
		} else if b == one {
			return 1
		} else if b == two {
			return 2
		} else if b == three {
			return 3
		} else if b == four {
			return 4
		} else if b == five {
			return 5
		} else if b == six {
			return 6
		} else if b == seven {
			return 7
		} else if b == eight {
			return 8
		} else if b == nine {
			return 9
		} else {
			panic("?")
		}
	}
	return 1000*digit(entry[10]) + 100*digit(entry[11]) + 10*digit(entry[12]) + digit(entry[13])
}

func (aoc *aoc202108) Part2(input *Input) string {
	in := make(chan [14]byte)
	out := make(chan int)
	for range runtime.NumCPU() {
		go func() {
			result := 0
			for entry := range in {
				result += aoc.decode(entry)
			}
			out <- result
		}()
	}
	for entry := range aoc.parse(input) {
		in <- entry
	}
	close(in)
	result := 0
	for range runtime.NumCPU() {
		result += <-out
	}
	return IntResult(result)
}
